Adding text format serialization / deserialization

BradLarson opened the pull request:
Per issue #36, this adds protobuf text format serialization and deserialization. To load a text format protobuf into a message, you use

```
let message = try Message(text:protobufText)
```

and to serialize it out to text, you use

```
try message.serializeText()
```

I'll admit, this is a bit of a hack, in that I copied over the JSON serialization / deserialization files and modified those until they matched the protobuf text format. A ton of code is therefore replicated between these files and could be abstracted out to a shared set of text/JSON input and output code.

My unit tests are rudimentary, but they exercise the data types from the Proto3TestAllTypes class, which seem to cover the available data types. I'm sure that I'm not handling all error cases correctly. 

I've tested this against a series of complex .prototxt text format files from the Caffe framework, and all of them appear to load correctly.

tbkka commented on Sources/SwiftProtobuf/ProtobufTextEncoding.swift:
```diff
@@ -0,0 +1,291 @@
+            append(text: value ? "true" : "false")
+        }
+    }
+    mutating func putStringValue(value: String) {
```
This was copied from the JSON support and uses JSON conventions.   Does protobuf text use `\u` escaping, or does it use UTF-8,  or...?

allevato replied:
The C++ code calls this `CEscapeAndAppend` function, which prints a C-escaped UTF-8 version of the string: https://github.com/google/protobuf/blob/fd046f6263fb17383cafdbb25c361e3451c31105/src/google/protobuf/stubs/strutil.cc#L579

tbkka commented on Sources/SwiftProtobuf/ProtobufTextEncoding.swift:
```diff
@@ -0,0 +1,291 @@
+            append(text: "\"NaN\"")
+        } else if !value.isFinite {
+            if value < 0 {
+                append(text: "\"-Infinity\"")
```
The "-Infinity" and similar cases need to be tested against Google's C++ implementation to make sure this is the correct coding.

tbkka commented on Sources/SwiftProtobuf/ProtobufTextEncoding.swift:
```diff
@@ -0,0 +1,291 @@
+        append(text: "}")
+    }
+    mutating func putNullValue() {
+        append(text: "null")
```
Does protobuf text encoding use "null"?

allevato replied:
No, `null` is only used by JSON format for certain special cases of the well-known types. None of those special cases apply for text format—WKTs are just serialized as ordinary messages with their fields.

tbkka commented on Sources/SwiftProtobuf/ProtobufTextDecoding.swift:
```diff
@@ -0,0 +1,1229 @@
+    case beginObject
+    case endObject
+    case beginArray
+    case endArray
```
Does protobuf text format use `[...]` array notation at all?

allevato replied:
Repeated fields can be repeated in text format, but there is also a shorthand bracket syntax: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L527

tbkka replied:
My #169 implements the bracket syntax.  It even handles that together with combining string fields and varying quotes, e.g.
```
some_field: ["abc", "def" 'ghi' "jkl", 'mno']
```
decodes to `["abc", "defghijkl", "mno"]`

tbkka commented on Sources/SwiftProtobuf/ProtobufTextDecoding.swift:
```diff
@@ -0,0 +1,1229 @@
+// "1e1000000".  The client code is responsible for real range
+// checking.
+//
+private func normalizeIntString(_ s: String) -> [Character]? {
```
This could likely be shared with JSON support, since these numeric conventions are very common.

allevato replied:
If we share them, we'll need to make sure that the numbers parsed by JSON format and text format are equivalent. Text format numeric parsing can be found starting here: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L900

tbkka replied:
Text format never codes integers using exponential notation, so this isn't needed there.

tbkka replied:
In #169, I completely rebuilt the numeric handling.  In particular, this was needed to correctly deal with octal and hexadecimal integer formats:  text format allows them for integer fields, rejects them for floating-point fields.

tbkka commented on Sources/SwiftProtobuf/ProtobufTextDecoding.swift:
```diff
@@ -0,0 +1,1229 @@
+        }
+    }
+    
+    mutating func decodeRepeatedGroupField<G: ProtobufMessage>(fieldType: G.Type, value: inout [G]) throws -> Bool {
```
Does protobufText handle groups differently from messages?  If not, this can simply defer to `decodeRepeatedMessageField()`

allevato replied:
There's a subtle difference with regard to name capitalization: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L445

tbkka replied:
I addressed this in #169 by changing the name that protoc-gen-swift puts into the fieldNames list.  If this is a group field, it now puts the group type name instead of the field name.  This matches the behavior of C++ protobuf text format.

I'm not sure why protoc puts a different field name into the descriptor than the name used by the text encoding.  That's a very odd mismatch.

allevato commented on Sources/SwiftProtobuf/ProtobufTextEncoding.swift:
```diff
@@ -0,0 +1,291 @@
+    
+    mutating public func visitUnknown(bytes: [UInt8]) {
+        // JSON encoding has no provision for carrying proto2 unknown fields
+        // TODO: Does this apply to the text format, as well?
```
Text format supports unknown fields (rather, the printer has a flag that lets you control whether they're printed or not). They get printed with their field number: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L1928

tbkka replied:
I've not implemented unknown field support in #169 yet.

allevato commented on Tests/SwiftProtobufTests/Helpers.swift:
```diff
@@ -104,6 +104,29 @@ extension PBTestHelpers where MessageTestType: ProtobufMessage & Equatable {
+            XCTAssert(expected == encoded, "Did not encode correctly: got \(encoded)", file: file, line: line)
+            do {
+                let decoded = try MessageTestType(text: encoded)
+//                print("==== Begin Decoding ====")
```
Delete commented out debugging code.

BradLarson commented:
Thanks for the comments. I'll see if I can clean up the easier elements first (mostly copy-paste items left over from the JSON parsing code), then move on to expansion of the tests. The JSON tests and some of the commented code are artifacts left over from my experimentation, and I simply forgot to pull them out when I reviewed this earlier.

This was just the minimum viable implementation to do what I needed, so there's a lot to clean up inside it. I'll see what I can do to improve it per these suggestions.

BradLarson commented:
Sorry it took me so long to respond, I had a few other things come up in the meantime.  I've updated the pull request per many of the review comments above:

- Renamed the Test_Text.swift file to Text_Text_proto3.swift to give room for proto2 tests.
- Removed the leftover JSON tests that I'd been using for debugging purposes.
- Split the text encoding tests into one per field type, and added some failure cases.
- Added tests for repeated items.
- Shorthand bracket syntax for repeated fields is supported, and added a test for that.
- Removed identified commented-out code.
- Removed left-over debugging statements.
- Removed unnecessary (_) pattern matches in case statements. 

These are issues that I haven't addressed yet, due to the limited time I've had to work on this:

- Failure test cases are not exhaustive.
- Not sure what to do about the errors that refer to JSON error cases.
- Not sure if I should combine the JSON and text numeric parsers, haven't dug into this: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L900
- Haven't addressed the differences in group name capitalization from here: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L445
- Don't handle text format unknown fields at present: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L1928
- I don't verify string escaping per here: https://github.com/google/protobuf/blob/fd046f6263fb17383cafdbb25c361e3451c31105/src/google/protobuf/stubs/strutil.cc#L579
- The "-Infinity" and similar cases need to be tested against Google's C++ implementation to make sure this is the correct coding.

If these are still blockers, I can keep working on the remaining issues.

tbkka commented:
I looked through the C++ unit tests some and have a few thoughts:

* I see no short-term need for duplicating the excessive number of formatting options supported by the C++ version.  Our encoder does not need to support UTF-8 and non-UTF8, "short" vs. "long" vs. "single line mode", configurable indent level, etc.

* Our text format decoder probably should support all of these, of course.  The C++ unit tests give a lot of examples we can use to build our own unit tests.

* Especially given that our unknown field support for protobuf binary is still evolving, I don't feel an immediate need to handle that for protobuf text format.

thomasvl commented:
For ObjC TextFormat generation I did utf8 only (Apple's tool chains generally do better with utf so there didn't seem to be a need to do non utf8); I did support an optional initial indent, but that was in part because I reused TextFormat generation for `-description`.

If we are going to support TextFormat parsing, we do need to support parsing all forms as they could come from any other implementation (or be hand written).

tbkka commented:
@BradLarson:  I just sent you a PR with a number of new tests that I cribbed from the C++ unit test suite.  Let me know if this is helpful...

tbkka commented:
I'm going to reconcile this with #139 today and hopefully fix some of the other outstanding issues with it.

tbkka commented:
@BradLarson:  I just pushed a bunch of work into my branch of this.  In particular:

* I restructured everything to match the new file organization (most files define just one type, filenames mostly match the type of that name, many of the `Protobuf` prefixes are gone)
* I rewrote the token struct and most of the scanner to better match protobuf text conventions (e.g., protobuf text has both an 'identifier' literal and a 'string' literal)
* I copied a bunch more tests from the protoc C++ test suite (lots more to go!)

With the above, I think I have most of the repeated field handling, octal and hex numeric literals, and a number of other things working.  I haven't yet looked at groups, maps, extensions or the proto3 Well-Known Types.

Let me know what you think...

tbkka commented:
Oh, yeah:  I also got `bytes` encoding/decoding working pretty completely.

BradLarson commented:
@tbkka - My apologies on being slow on the response. Had a few things that drew me away at work. I'll see if I can't review all this tomorrow.

tbkka commented:
Although still incomplete, I think #169 is getting close.  Any feedback greatly appreciated.

BradLarson commented:
Given the work that tbkka's done on this over a #169, I'll close this out in favor of that pull request.
